### PR TITLE
ARROW-757: [C++] Resolve nmake build issues on Windows

### PR DESCRIPTION
**Pull request by maxhora**

### CODE REVIEW

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 1]
```diff
@@ -24,8 +24,11 @@ CHECK_CXX_COMPILER_FLAG("-msse3" CXX_SUPPORTS_SSE3)
 CHECK_CXX_COMPILER_FLAG("-maltivec" CXX_SUPPORTS_ALTIVEC)
 
 # compiler flags that are common across debug/release builds
-#  - Wall: Enable all warnings.
-set(CXX_COMMON_FLAGS "-std=c++11 -Wall")
+#  - Wall: Enable all warnings. Not used for MSVC,
+# because cause build failure by too much warnings.
+if (NOT MSVC)
+    set(CXX_COMMON_FLAGS "-std=c++11 -Wall")
```
Comment (xhochy) [inline cpp/cmake_modules/SetupCxxFlags.cmake]: Don't we need to explicitly set C++11 for MSVC?

Comment (MaxRis) [inline cpp/cmake_modules/SetupCxxFlags.cmake]: I have found that nmake complained that -std=c++11 command is not recognized. According to https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler/ "Note that there is no plan to add a C++11 switch.". It seems MS recently added /std:c++14 switch to VS 2015 nmake distribution, but all previous C++ standards implemented features are enabled by default and there are no command line switches, etc. Pretty weird.

**Comment by wesm** [comment]
I tried running this locally (VS Community 2017) but got the following error:

```
[ 24%] Building CXX object CMakeFiles/arrow_objlib.dir/src/arrow/util/bit-util.cc.obj
bit-util.cc
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\include\xlocale(314): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
C:\Users\wesm\code\arrow\cpp\src\arrow\util\bit-util.cc(81): error C3861: '_mm_popcnt_u64': identifier not found
NMAKE : fatal error U1077: 'C:\PROGRA~2\MICROS~2\2017\COMMUN~1\VC\Tools\MSVC\1410~1.250\bin\HostX86\x86\cl.exe' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\bin\HostX86\x86\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\bin\HostX86\x86\nmake.exe"' : return code '0x2'
Stop.
```

This is the output of

```
cmake .. -G "NMake Makefiles"
nmake
```

like

```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\include\xlocale(314): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.10.25017\include\exception(366): warning C4577: 'noexcept' used with no exception handling mode specified; termination on exception is not guaranteed. Specify /EHsc
```

Aside, there's a lot of warnings when compiling, even without `/Wall`. We might want to go through http://llvm.org/viewvc/llvm-project/llvm/trunk/cmake/modules/HandleLLVMOptions.cmake?view=markup from the LLVM project and build up a similar set of explicit warning suppressions for MSVC

**Comment by wesm** [comment]
This compiles with 

```
cmake -G "Visual Studio 14 2015 Win64" -DARROW_BOOST_USE_SHARED=OFF -DARROW_BUILD_BENCHMARKS=OFF -DARROW_JEMALLOC=OFF -DCMAKE_BUILD_TYPE=Release ..
cmake --build . --config Release
```

though

**Comment by MaxRis** [comment]
Modified changes to remove -Wall flag only for gtest subproject building with MSVC.

**Comment by MaxRis** [comment]
Builds locally fine for me with command line:

cmake -G "NMake Makefiles" -DARROW_BOOST_USE_SHARED=ON -DARROW_BUILD_BENCHMARKS=OFF -DARROW_JEMALLOC=OFF -DCMAKE_BUILD_TYPE=Release ..

**Comment by wesm** [comment]
The correct command to launch the command prompt for amd64 architecture on VS 2017 Community is

```
 "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\Tools\VsDevCmd.bat" -arch=amd64
```

I will write a patch to provide developer instructions

**Pull request closed without merging**